### PR TITLE
ONNX export for custom architectures & models with custom modeling code

### PR DESCRIPTION
**Pull request by fxmarty**
As per title. As a next step, we should support loading an `onnx_config.py` file from the custom model repository and use it for the ONNX export, so that e.g. the CLI works as well.

Fixes https://github.com/huggingface/optimum/issues/1061 https://github.com/huggingface/optimum/issues/1040 https://github.com/huggingface/optimum/issues/1134

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "torchvision",
     "diffusers>=0.17.0",
     "torchaudio",
+    "einops",
```
Comment (michaelbenayoun) [inline setup.py:36]: Why?

**Review by michaelbenayoun — APPROVED** [review]
LGTM

File: setup.py (hunk 1 above)
Comment (fxmarty) [reply setup.py:36]: MPT custom code requires it

**Review by echarlaix — APPROVED** [review]
It looks great @fxmarty

**Comment by fxmarty** [comment]
@zacharyblank Are you using the correct branch? The tests does pass

**Comment by zacharyblank** [comment]
@fxmarty ran with `pip install git+https://github.com/fxmarty/optimum@extended-custom-onnx-export`

Copied the code in the docs exactly: https://github.com/huggingface/optimum/pull/1166/files#diff-d9f4133ab6304f52662dca9bd9b2a526f9b8d5e54898c4ba35e4c023e28f4c75R323

**Comment by zacharyblank** [comment]
@fxmarty disregard. My mistake. pip fooled me again.

pip: 1
zach: 0

**Comment by zacharyblank** [comment]
@fxmarty I was able to convert MPT to onnx using the code and example in the PR you created. But, I am now running into inference issues. Seems that I need to fix the length in order for it to work™ but with text generation I need a dynamic length. I understand that with ONNX I need a static sequence length but I don't understand how the works with text generation, new tokens, truncation and padding.

For example. My code:

```
model_path = "/home/paperspace/Isomeric/mpt_onnx"
config = AutoConfig.from_pretrained(model_path, trust_remote_code=True)
model = ORTModelForCausalLM.from_pretrained(model_path, 
                                            config=config, 
                                            provider="CUDAExecutionProvider"
                                            )

tokenizer = AutoTokenizer.from_pretrained(model_path)

inputs = tokenizer("This is only a test and if you", 
                   return_tensors="pt", 
                   truncation=True, 
                   padding="max_length", 
                   max_length=128,
                ).to("cuda")

gen_tokens = model.generate(**inputs,
                            do_sample=True,
                            temperature=0.1,
                            max_length=129,
                        )

tokenizer.batch_decode(gen_tokens)
```

Produces this output:

```
["This is only a test and if you<|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|>'t"]
```

If I don't set `max_length` `padding` and `truncation` as such I get this error:

```
RuntimeError: Error in execution: Non-zero status code returned while running Concat node. 
Name:'/transformer/blocks.0/attn/Concat_6' Status Message: 
/onnxruntime_src/onnxruntime/core/framework/execution_frame.cc:171 onnxruntime::common::Status 
onnxruntime::IExecutionFrame::GetOrCreateNodeOutputMLValue(int, int, const onnxruntime::TensorShape*, OrtValue*&, 
const onnxruntime::Node&) shape && tensor.Shape() == *shape was false. OrtValue shape verification failed. Current 
shape:{1,32,128,129} Requested shape:{1,32,128,9}
```

**Pull request merged**